engine/collection/compliance/core.go

```diff
@@ -200,21 +205,50 @@ func (c *Core) OnBlockProposal(originID flow.Identifier, proposal *messages.Clus
 func (c *Core) processBlockAndDescendants(proposal *messages.ClusterBlockProposal) error {
 	blockID := proposal.Header.ID()
 
-	// process block itself
-	err := c.processBlockProposal(proposal)
-	// child is outdated by the time we started processing it
-	// => node was probably behind and is catching up. Log as warning
-	if engine.IsOutdatedInputError(err) {
-		c.log.Info().Msg("dropped processing of abandoned fork; this might be an indicator that the node is slightly behind")
-		return nil
-	}
-	// the block is invalid; log as error as we desire honest participation
-	// ToDo: potential slashing
-	if engine.IsInvalidInputError(err) {
-		c.log.Warn().Err(err).Msg("received invalid block from other node (potential slashing evidence?)")
-		return nil
+	// retrieve the parent block
+	//  - once we reach this point, the parent block must have been validated
+	//    and inserted to the protocol state
+	parent, err := c.headers.ByBlockID(proposal.Header.ParentID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve proposal parent: %w", err)
 	}
+
+	log := c.log.With().
+		Str("block_id", blockID.String()).
+		Uint64("block_height", proposal.Header.Height).
+		Uint64("block_view", proposal.Header.View).
+		Uint64("parent_view", parent.View).
+		Logger()
+
+	// process block itself
+	err = c.processBlockProposal(proposal, parent)
 	if err != nil {
+		if engine.IsOutdatedInputError(err) {
+			// child is outdated by the time we started processing it
+			// => node was probably behind and is catching up. Log as warning
+			log.Info().Msg("dropped processing of abandoned fork; this might be an indicator that the node is slightly behind")
+			return nil
+		} else if engine.IsInvalidInputError(err) {
+			// the block is invalid; log as error as we desire honest participation
+			// ToDo: potential slashing
+			log.Warn().Err(err).Msg("received invalid block from other node (potential slashing evidence?)")
+			// notify the VoteAggregator of an invalid block to enable detecting Byzantine voters
+			err = c.voteAggregator.InvalidBlock(model.ProposalFromFlow(proposal.Header, parent.View))
+			if err != nil {
+				if mempool.IsBelowPrunedThresholdError(err) {
+					log.Warn().Msg("received invalid block, but is below pruned threshold")
+					return nil
+				} else {
+					return fmt.Errorf("unexpected error notifying vote aggregator about invalid block: %w", err)
+				}
+			}
+			return nil
+		} else if engine.IsUnverifiableInputError(err) {
```

Comment:
I believe this is also slashable, malicious replicas could have voted for this proposal, need to notify VA

Reply:
Right, to actually submit a slashing challenge we would need slightly more information from elsewhere, in particular a confirmation of a finalized block before the epoch commitment deadline, but it is knowable. Agree we should notify VA for this case.